experimental/codepointtrie/src/codepointtrie.rs

```diff
@@ -0,0 +1,258 @@
+        // dynamic dispatch here.
+
+        let trie: CodePointTrie<'trie, W, T> = CodePointTrie {
+            header,
+            index,
+            data,
+            _marker_ty: PhantomData,
+        };
+        Ok(trie)
+    }
+
+    fn trie_internal_small_index(&self, code_point: u32) -> u32 {
+        let mut index1_pos: u32 = code_point >> SHIFT_1;
+        if T::ENUM_VALUE == TrieTypeEnum::Fast {
+            assert!(
+                FAST_TYPE_FAST_INDEXING_MAX < code_point && code_point < self.header.high_start
+            );
+            index1_pos = index1_pos + BMP_INDEX_LENGTH - OMITTED_BMP_INDEX_1_LENGTH;
+        } else {
+            assert!(code_point < self.header.high_start && self.header.high_start > SMALL_LIMIT);
+            index1_pos += SMALL_INDEX_LENGTH;
+        }
+        let index3_block_idx: u32 = (self.index.get(index1_pos as usize).unwrap() as u32)
+            + ((code_point >> SHIFT_2) & INDEX_2_MASK);
+        let mut index3_block: u32 = self.index.get(index3_block_idx as usize).unwrap() as u32;
+        let mut index3_pos: u32 = (code_point >> SHIFT_3) & INDEX_3_MASK;
+        let mut data_block: u32;
+        if index3_block & 0x8000 == 0 {
+            // 16-bit indexes
+            data_block = self
+                .index
+                .get((index3_block + index3_pos) as usize)
+                .unwrap() as u32;
+        } else {
+            // 18-bit indexes stored in groups of 9 entries per 8 indexes.
+            index3_block = (index3_block & 0x7fff) + (index3_pos & !7) + (index3_pos >> 3);
+            index3_pos &= 7;
+            data_block = ((self.index.get(index3_block as usize).unwrap() << (2 + (2 * index3_pos)))
+                as u32
+                & 0x30000) as u32;
+            index3_block += 1;
+            data_block |= self
+                .index
+                .get((index3_block + index3_pos) as usize)
+                .unwrap() as u32;
+        }
+        // Return data_pos == data_block (offset) +
+        //     portion of code_point bit field for last (4th) lookup
+        data_block + (code_point & SMALL_DATA_MASK)
+    }
+
+    /// Internal trie getter for a code point at or above the fast limit that
+    /// is designed for this trie's trie type, [`T`]. Returns the position in
+    /// the `data` array for the value that is associated with `code_point`.
+    fn trie_above_fastmax_index(&self, code_point: u32) -> u32 {
+        if code_point >= self.header.high_start {
+            self.header.data_length - HIGH_VALUE_NEG_DATA_OFFSET
+        } else {
+            self.trie_internal_small_index(code_point) // helper fn
+        }
+    }
+
+    /// Internal trie getter for a code point below the fast limit that
+    /// is designed for this trie's trie type, [`T`]. Returns the position in
+    /// the `data` array for the value that is associated with `code_point`.
+    fn trie_below_fastmax_index(&self, c: u32) -> u32 {
```

Comment:
It would be nice to use the same method names as in C++ & Java, modulo CamelCasing vs. underscores etc.

Reply:
I had review comments to rename identifiers and functions to be more readable, even while acknowledging it was ported code. In my mind, it seemed that the thing that tripped people up was the re-use of the words "small" and "fast" as both a TrieType (Small x Fast) and the description of the regions of the index array below and above the fastMax value that separates the 2-stage lookups from the 4-stage lookups ("fast index" x "small index"), even though those are orthogonal concepts. OTOH, I'm not sure whether the renaming of local variable identifiers inside the functions to implement `.get()` helped or made things more confusing. I had trouble coming up with choosing more descriptive names in both places.

Reply:
I think this project is a great opportunity to get more eyes on this code than we had when Markus implemented it in ICU4C.  I think we should use names that make sense for the reviewers here, and then perhaps migrate the name changes *back to ICU4C and ICU4J*.

Reply:
I don't see any big naming problem in the C/Java code. Yes, we have fast & small tries, and fast-index-lookup vs. small-index-lookup. The point of the type-fast version is that fastIndex() is called for a larger range of code points than in a type-small version -- that makes it faster for more code points.

Reply:
2021-07-22: Elango to use the ICU4C names but add documentation explaining what they do and what the difference is between a fast trie and a fast-index-lookup.